src/lib.rs

```diff
@@ -732,36 +731,66 @@ impl<'a, R: Rng> Iterator for AsciiGenerator<'a, R> {
     }
 }
 
-/// A random number generator that can be explicitly seeded to produce
-/// the same stream of randomness multiple times.
-pub trait SeedableRng<Seed>: Rng {
-    /// Reseed an RNG with the given seed.
+mod private {
+    pub trait Sealed {}
+    impl Sealed for [u8; 4] {}
+    impl Sealed for [u8; 8] {}
+    impl Sealed for [u8; 12] {}
+    impl Sealed for [u8; 16] {}
+    impl Sealed for [u8; 24] {}
+    impl Sealed for [u8; 32] {}
+}
+
+/// The seed type is restricted to these types. This trait is sealed to prevent
+/// user-extension.
+///
+/// Use of byte-arrays avoids endianness issues. We may extend this to allow
+/// byte arrays of other lengths in the future.
+pub trait SeedRestriction: private::Sealed + Default + AsMut<[u8]> {}
+impl<S> SeedRestriction for S where S: private::Sealed + Default + AsMut<[u8]> {}
+
+/// A random number generator that can be explicitly seeded.
+///
+/// There are two subtle differences between `from_rng` and`from_seed` (beyond
+/// the obvious): first, that `from_rng` has no reproducibility requirement, and
+/// second, that `from_rng` may directly fill internal states larger than
+/// `SeedableRng::Seed`, where `from_seed` may need some extra step to expand
+/// the input.
+pub trait SeedableRng: Sized {
+    /// Seed type.
+    type Seed: SeedRestriction;
+
+    /// Create a new PRNG using the given seed.
     ///
-    /// # Example
+    /// Each PRNG should implement this.
     ///
-    /// ```rust
-    /// use rand::{Rng, SeedableRng, StdRng};
+    /// Reproducibility is required; that is, a fixed PRNG seeded using this
+    /// function with a fixed seed should produce the same sequence of output
+    /// today, and in the future. PRNGs not able to satisfy this should make
+    /// clear notes in their documentation. It is however not required that this
+    /// function yield the same state as a reference implementation of the PRNG
+    /// given equivalent seed; if necessary another constructor should be used.
     ///
-    /// let seed: &[_] = &[1, 2, 3, 4];
-    /// let mut rng: StdRng = SeedableRng::from_seed(seed);
-    /// println!("{}", rng.gen::<f64>());
-    /// rng.reseed(&[5, 6, 7, 8]);
-    /// println!("{}", rng.gen::<f64>());
-    /// ```
```

Comment:
I have removed the examples for now. When we split this off into the `rand_core` crate, we can't run tests here anyway because there are no PRNG's available?